posts/benchmarking_design_decisions/index.qmd

```diff
@@ -0,0 +1,246 @@
+---
+title: "Using benchmarking to guide the adoption of dependencies in R packages"
+author:
+  - name: "James Mba Azam"
+    orcid: "https://orcid.org/0000-0001-5782-7330"
+date: last-modified
+categories: [benchmarking, R, tidyverse, cli, R packages, software architecture]
+format:
+  html:
+    toc: true
+execute: 
+  cache: true
+---
+
+```{r, echo = FALSE}
+knitr::opts_chunk$set(
+  cache = TRUE,
+  message = FALSE
+)
+```
+
+R package developers are constantly faced with design decisions about user
+friendliness. Such design decisions often include checking that inputs are
+rightly specified and returning feedback early, deciding on how to deal with
+issues arising during code execution within a function, and how to
+structure outputs.
+
+These design decisions often result in taking on dependencies. Dependencies are
+external packages that enable a function to work properly. For example, the
+`tidyverse` is a collection of packages that are often used together to
+make data analysis easier. The `tidyverse` is a dependency for many
+packages that are built on top of it.
+
+Dependencies should only be adopted after carefully considering their pros and
+cons. In this article, we will demonstrate how to use (micro)benchmarking[^1]
+to guide the decision to take on dependencies that might lead to slower
+processes. In data analytics pipelines, speed is often a key consideration,
+especially when working with large datasets. Therefore, it is important to
+consider how the new dependency might impact the speed of an existing
+process.
+
+For this demonstration, we will look at condition signalling,
+that is, how to signal to a user that an issue has arisen during code
+execution [^2]. Condition signalling in R is classified into errors,
+warnings, and messages. Errors refer
+to cases where something fatal happens during code execution so that it halts,
+for example, when a function is run without specifying the required arguments.
+
+```{r}
+try(rnorm())
+```
+
+Warnings are used to handle less fatal issues with code execution that do not
+need to halt the process. For example, when you compare two vectors of
+different lengths, R will throw a warning but will still return a result.
+
+```{r}
+c(1:2) > c(1:3)
+```
+
+Messages are used to provide useful information about processes. For
+example, when you load a package, R will throw a message to indicate that the
+package has been loaded.
```

Comment:
I think this only happens when there are namespace conflicts, and otherwise the package loads silently (assuming the package author didn't write a message to be printed in `.onLoad()`).

Reply:
True. I'll make the example more specific to this package.

Reply:
Thanks @pratikunterwegs. I'll add the qmd style option in the chunk.